Zend/tests/bugGH-8655.phpt

```diff
@@ -0,0 +1,29 @@
+--TEST--
+Bug GH-8655 (zval reference is not released when targetting a declared property)
+--FILE--
+<?php
+class Foo
+{
+    public $foo;
+}
+
+function hydrate($properties, $object)
+{
+    foreach ($properties as $name => &$value) {
+        $object->$name = &$value;
+    }
+};
+
+$object = new Foo;
+
+hydrate(['foo' => 123], $object);
+
+$arrayCast = (array) $object;
+
+$object->foo = 234;
+var_dump(ReflectionReference::fromArrayElement($arrayCast, 'foo'));
+echo $arrayCast['foo'];
+?>
+--EXPECT--
+NULL
+123
```

Comment:
```suggestion
$object = new Foo;
hydrate(['foo' => 123], $object);
$arrayCast = get_object_vars($object);

$object->foo = 234;
var_dump(ReflectionReference::fromArrayElement($arrayCast, 'foo'));
var_dump($arrayCast['foo']);

// casting object to array must work the same as get_object_vars()
$object = new Foo;
hydrate(['foo' => 125], $object);
$arrayCast = (array) $object;

$object->foo = 236;
var_dump(ReflectionReference::fromArrayElement($arrayCast, 'foo'));
var_dump($arrayCast['foo']);
?>
--EXPECT--
NULL
int(123)
NULL
int(125)
```

this is definitely a bug, `get_object_vars()` works correctly in PHP 8.1 - https://3v4l.org/hg1Do

out of curiosity, why is `get_object_vars()` not affected, does this function behave differently than the cast to an array?

Reply:
This function and test is specifically about array casts. get_object_vars is unrelated. Also no, get_object_vars() and array casts generally are not identical.
E.g. `get_object_vars` is doing property permission checks as well as unmangling, which array casts are not doing.